// @alethio/ui/lib/control/accordion/internal/AccordionState.js
import * as tslib_1 from "tslib";
import { observable, action } from "mobx";
import { AccordionItemState } from "../AccordionItemState";
import { AccordionItemContentStatus } from "../AccordionItemContentStatus";
export class AccordionState {
    constructor(onError) {
        this.onError = onError;
        this.handleItemClick = (item) => tslib_1.__awaiter(this, void 0, void 0, function* () {
            this.activeItem = item === this.activeItem ? void 0 : item;
            if (this.activeItem !== void 0) {
                yield this.onItemExpanded(item);
            }
        });
        this.items = [];
    }
    getItems() {
        return this.items;
    }
    getActiveItem() {
        return this.activeItem;
    }
    createItem(itemConfig, idx) {
        let item = new AccordionItemState();
        item.index = idx;
        item.config = itemConfig;
        item.onClick = () => this.handleItemClick(item);
        return item;
    }
    addItem(itemConfig) {
        this.items.push(this.createItem(itemConfig, this.items.length));
    }
    removeItem(itemConfig) {
        let itemIndex = this.items.findIndex(it => it.config === itemConfig);
        if (itemIndex !== -1) {
            if (this.activeItem === this.items[itemIndex]) {
                this.activeItem = void 0;
            }
            this.items.splice(itemIndex, 1);
        }
    }
    updateItem(oldConfig, newConfig) {
        let item = this.items.find(it => it.config === oldConfig);
        if (item) {
            if (item.config.content !== newConfig.content && this.activeItem === item) {
                this.onItemExpanded(item);
            }
            item.config = newConfig;
        }
    }
    onItemExpanded(item) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            try {
                item.resetContent();
                item.updateContent(yield item.config.content(), AccordionItemContentStatus.Loaded);
            }
            catch (e) {
                this.onError(e, item);
                item.updateContent(void 0, AccordionItemContentStatus.Error);
            }
        });
    }
}
tslib_1.__decorate([
    observable.shallow
], AccordionState.prototype, "items", void 0);
tslib_1.__decorate([
    observable.ref
], AccordionState.prototype, "activeItem", void 0);
tslib_1.__decorate([
    action
], AccordionState.prototype, "removeItem", null);
tslib_1.__decorate([
    action
], AccordionState.prototype, "updateItem", null);
