// @alethio/explorer-ui/lib/dashboard/BarChartItem.js
import * as React from "react";
import styled from "@alethio/ui/lib/styled-components";
import { Tooltip } from "@alethio/ui/lib/overlay/tooltip/Tooltip";
import { Link } from "plugin-api/component/Link";
import { BarChartItemBar } from "./BarChartItemBar";
export const BarChartItemRoot = styled.div `
    cursor: pointer;
`;
const BarChartItemHitBox = styled("div") `
    height: ${({ maxHeight }) => maxHeight}px;
    padding-left: 4px;
    padding-right: 4px;
    display: flex;
    align-items: flex-end;
`;
const BarChartItemBarWrapper = styled("div") `
    width: 8px;
    height: ${props => Math.round(props.percent * props.maxHeight / 100)}px;
    min-height: 8px;
`;
export class BarChartItem extends React.Component {
    render() {
        return (React.createElement(BarChartItemRoot, null,
            React.createElement(Tooltip, { placement: "top", nonInteractive: true, showDelay: 0, hideDelay: 0, offset: 10, referenceElement: () => this.barEl, content: this.props.tooltip },
                React.createElement(Link, { to: this.props.link },
                    React.createElement(BarChartItemHitBox, { maxHeight: this.props.maxHeight },
                        React.createElement(BarChartItemBarWrapper, { percent: this.props.value, maxHeight: this.props.maxHeight },
                            React.createElement(BarChartItemBar, { active: this.props.active, innerRef: ref => this.barEl = ref })))))));
    }
}
