// @alethio/ui/lib/layout/topbar/HorizontalBar.js
import * as tslib_1 from "tslib";
import * as React from "react";
import styled, { css } from "../../styled-components";
import { Bar } from "../Bar";
const Placeholder = styled("div").withConfig({ displayName: "Placeholder", componentId: "sc-10cjr3d" }) `
    height: ${props => props.height}px;
    width: 100%;

    /**
     * TODO: Responsiveness css does not belong here, but because we must also put the css into the placeholder
     * we leave it here for now
     */
    display: none;
    @media ${props => props.theme.mediaQueries.breakPoints.smallerThanStandardView} {
        display: block;
    }
`;
const HorizontalBarRoot = styled(Bar).withConfig({ displayName: "HorizontalBarRoot", componentId: "sc-11ft1qt" }) `
    height: ${props => props.height}px;

    width: 100%;
    ${props => props.sticky ? css `
    position: fixed;
    top: 0;
    ` : ""}
    border-bottom: 1px solid ${props => props.theme.colors.sidebarBorder};

    /**
     * TODO: Responsiveness css does not belong here, but because we must also put the css into the placeholder
     * we leave it here for now
     */
    display: none;
    @media ${props => props.theme.mediaQueries.breakPoints.smallerThanStandardView} {
        display: flex;
    }
`;
export class HorizontalBar extends React.Component {
    render() {
        let _a = this.props, { contentRef } = _a, props = tslib_1.__rest(_a, ["contentRef"]);
        return (React.createElement(React.Fragment, null,
            this.props.sticky ? React.createElement(Placeholder, { height: this.props.height }) : null,
            React.createElement(HorizontalBarRoot, Object.assign({ innerRef: contentRef }, props))));
    }
}
