// @alethio/explorer-ui/lib/blockTxs/TxsViewMode.js
import * as tslib_1 from "tslib";
import { observable, computed, action } from "mobx";
var ViewModes;
(function (ViewModes) {
    ViewModes[ViewModes["TableList"] = 0] = "TableList";
    ViewModes[ViewModes["HeatMapGrid"] = 1] = "HeatMapGrid";
})(ViewModes || (ViewModes = {}));
export class TxsViewMode {
    constructor() {
        this.txsViewMode = 1;
    }
    get getTxsViewMode() {
        return this.txsViewMode;
    }
    setTxsViewMode(mode) {
        this.txsViewMode = mode;
    }
    setToTableList() {
        this.setTxsViewMode(0);
    }
    setToHeatMapGrid() {
        this.setTxsViewMode(1);
    }
    get isTableList() {
        return this.txsViewMode === 0;
    }
    get isHeatMapGrid() {
        return this.txsViewMode === 1;
    }
}
tslib_1.__decorate([
    observable
], TxsViewMode.prototype, "txsViewMode", void 0);
tslib_1.__decorate([
    computed
], TxsViewMode.prototype, "getTxsViewMode", null);
tslib_1.__decorate([
    action
], TxsViewMode.prototype, "setTxsViewMode", null);
tslib_1.__decorate([
    computed
], TxsViewMode.prototype, "isTableList", null);
tslib_1.__decorate([
    computed
], TxsViewMode.prototype, "isHeatMapGrid", null);
