// @alethio/explorer-ui/lib/blockTxs/ViewSelectorBox.js
import * as tslib_1 from "tslib";
import styled from "@alethio/ui/lib/styled-components";
import * as React from "react";
import { observer } from "mobx-react";
import { ViewSelectorButton } from "./ViewSelectorButton";
import { ListViewIcon } from "@alethio/ui/lib/icon/ListViewIcon";
import { GridViewIcon } from "@alethio/ui/lib/icon/GridViewIcon";
const HEIGHT = 28;
const ViewSelectorRoot = styled.div `
    padding: 0 7px;
    display: flex;
    width: ${2 * HEIGHT + 24}px;
    justify-content: space-between;
`;
let ViewSelectorBox = class ViewSelectorBox extends React.Component {
    constructor() {
        super(...arguments);
        this.onListClick = () => {
            this.props.txsViewMode.setToTableList();
        };
        this.onGridClick = () => {
            this.props.txsViewMode.setToHeatMapGrid();
        };
    }
    render() {
        return (React.createElement(ViewSelectorRoot, null,
            React.createElement(ViewSelectorButton, { active: this.props.txsViewMode.isTableList, onClick: this.onListClick },
                React.createElement(ListViewIcon, null)),
            React.createElement(ViewSelectorButton, { active: this.props.txsViewMode.isHeatMapGrid, onClick: this.onGridClick },
                React.createElement(GridViewIcon, null))));
    }
};
ViewSelectorBox = tslib_1.__decorate([
    observer
], ViewSelectorBox);
export { ViewSelectorBox };
