// @alethio/explorer-ui/lib/grid/dataRenderer/tx/TransferDirectionRenderer.js
import * as React from "react";
import styled from "@alethio/ui/lib/styled-components";
import { MsgCountOutSmallIcon } from "@alethio/ui/lib/icon/MsgCountOutSmallIcon";
import { MsgCountInSmallIcon } from "@alethio/ui/lib/icon/MsgCountInSmallIcon";
import { TooltipRegular } from "@alethio/ui/lib/overlay/tooltip/TooltipRegular";
let ICON_SIZE = 24;
const DirectionWrapper = styled.div `
    position: relative;
    margin-top: -2px;

    > * {
        position: absolute;
        top: 0;
        left: 0;
        transform: translate(-50%, -50%);
    }
`;
export class TransferDirectionRenderer {
    constructor(accountAddress, tr) {
        this.accountAddress = accountAddress;
        this.tr = tr;
    }
    render(f) {
        let directionOut = f.from === this.accountAddress;
        return React.createElement(DirectionWrapper, null, directionOut ?
            React.createElement(TooltipRegular, { content: this.tr.get("accountView.content.transferDirection.outgoing.tooltip") },
                React.createElement(MsgCountOutSmallIcon, { size: ICON_SIZE }))
            :
                React.createElement(TooltipRegular, { content: this.tr.get("accountView.content.transferDirection.incoming.tooltip") },
                    React.createElement(MsgCountInSmallIcon, { size: ICON_SIZE })));
    }
}
