// @alethio/ui/lib/overlay/tooltip/Tooltip.js
import * as tslib_1 from "tslib";
import * as React from "react";
import { observer } from "mobx-react";
import { observable } from "mobx";
import { Popover } from "../Popover";
import { DomNodeProxy } from "../../util/react/DomNodeProxy";
import { MultiHover } from "../../util/react/MultiHover";
let Tooltip = class Tooltip extends React.Component {
    constructor(props) {
        super(props);
        this.visible = false;
        this.handleMouseEnter = () => {
            this.multiHover.enter(true);
        };
        this.handleMouseLeave = () => {
            this.multiHover.leave();
        };
        this.multiHover = new MultiHover({ enterDelay: this.props.showDelay, leaveDelay: this.props.hideDelay });
        this.multiHover.onActiveChange.subscribe(visible => this.visible = visible || false);
    }
    componentDidUpdate(prevProps) {
        if (this.props.showDelay !== prevProps.showDelay) {
            this.multiHover.applyOpts({ enterDelay: this.props.showDelay });
        }
        if (this.props.hideDelay !== prevProps.hideDelay) {
            this.multiHover.applyOpts({ leaveDelay: this.props.hideDelay });
        }
    }
    render() {
        let _a = this.props, { showDelay: _unused, hideDelay: _unused2, content } = _a, otherProps = tslib_1.__rest(_a, ["showDelay", "hideDelay", "content"]);
        let target = React.Children.only(this.props.children);
        return (React.createElement(Popover, Object.assign({ visible: this.visible, content: React.createElement("div", { onMouseEnter: this.handleMouseEnter, onMouseLeave: this.handleMouseLeave }, content) }, otherProps),
            React.createElement(DomNodeProxy, { onMount: el => {
                    el.addEventListener("mouseenter", this.handleMouseEnter);
                    el.addEventListener("mouseleave", this.handleMouseLeave);
                }, onUnmount: el => {
                    el.removeEventListener("mouseenter", this.handleMouseEnter);
                    el.removeEventListener("mouseleave", this.handleMouseLeave);
                } }, target)));
    }
};
Tooltip.defaultProps = {
    showDelay: 400,
    hideDelay: 400
};
tslib_1.__decorate([
    observable
], Tooltip.prototype, "visible", void 0);
Tooltip = tslib_1.__decorate([
    observer
], Tooltip);
export { Tooltip };
