// @alethio/ui/lib/control/Button.js
import * as React from "react";
import styled, { css } from "../styled-components";
import { Box } from "../layout/content/box/Box";
import { HoverState } from "../util/react/HoverState";
const ButtonRoot = styled("div").withConfig({ displayName: "ButtonRoot", componentId: "sc-y819sc" }) `
    cursor: pointer;
    text-transform: uppercase;
    ${props => props.floating ? css `
    box-shadow: 0 8px 16px rgba(167, 181, 209, 0.6);
    ` : ``}
    ${props => props.disabled ? css `
    opacity: 0.75;
    ` : void 0}
`;
const StyledBox = styled(Box).withConfig({ displayName: "StyledBox", componentId: "sc-3nfx7i" }) `
    border-radius: 4px;
    transition: background-color .2s ease-in-out, border-color .2s ease-in-out;
`;
const colorSets = {
    primary: (hover) => (theme) => ({
        background: hover ? theme.colors.buttonPrimaryBgActive : theme.colors.buttonPrimaryBg,
        text: theme.colors.buttonPrimaryText
    }),
    secondary: (hover) => (theme) => ({
        background: hover ? theme.colors.buttonSecondaryBgActive : theme.colors.buttonSecondaryBg,
        text: theme.colors.buttonSecondaryText
    })
};
export class Button extends React.Component {
    render() {
        let { Icon, iconPlacement, floating, disabled, colors, children } = this.props;
        return (React.createElement(HoverState, null, (hover) => React.createElement(ButtonRoot, { onClick: !this.props.disabled ? this.props.onClick : void 0, floating: floating, disabled: disabled },
            React.createElement(StyledBox, { Icon: Icon, iconPlacement: iconPlacement ? iconPlacement : "left", colors: colorSets[colors](!this.props.disabled ? hover : false), metrics: {
                    fontSize: 12,
                    lineHeight: 14,
                    fontWeight: 700,
                    height: 36,
                    iconSize: 24,
                    letterSpacing: .4,
                    textPaddingTop: 9,
                    textPaddingX: 16
                } }, children))));
    }
}
Button.defaultProps = {
    colors: "secondary"
};
