// @alethio/ui/lib/control/Checkbox.js
import * as React from "react";
import styled from "../styled-components";
import { CheckboxOnIcon } from "../icon/CheckboxOnIcon";
import { CheckboxOffIcon } from "../icon/CheckboxOffIcon";
const CheckboxWrapper = styled.div.withConfig({ displayName: "CheckboxWrapper", componentId: "sc-wqgt27" }) `
    padding: 8px 8px 8px 32px;
    position: relative;
`;
const CheckboxLabel = styled.label.withConfig({ displayName: "CheckboxLabel", componentId: "sc-1jryfg9" }) `
    text-transform: uppercase;
    display: block;
    font-size: 12px;
    line-height: 16px;
    font-weight: 600;
    color: ${({ theme }) => theme.colors.checkboxLabel};
`;
const CheckboxIconWrapper = styled.div.withConfig({ displayName: "CheckboxIconWrapper", componentId: "sc-cmbprq" }) `
    position: absolute;
    top: 4px;
    left: 4px;
    color: ${({ theme }) => theme.colors.checkboxIcon};
`;
export class Checkbox extends React.PureComponent {
    constructor() {
        super(...arguments);
        this.onChange = (event) => {
            if (this.props.onChange) {
                this.props.onChange(event, !this.props.checked, this.props.name, this.props.value);
            }
        };
    }
    render() {
        return (React.createElement(CheckboxWrapper, null,
            React.createElement("input", { type: "checkbox", id: this.props.id, checked: this.props.checked || false, value: this.props.value, onChange: this.onChange, style: { display: "none" } }),
            React.createElement(CheckboxLabel, { htmlFor: this.props.id },
                React.createElement(CheckboxIconWrapper, null, this.props.checked ? React.createElement(CheckboxOnIcon, null) : React.createElement(CheckboxOffIcon, null)),
                this.props.children)));
    }
}
