// @alethio/ui/lib/control/ModalSelectBox.js
import * as React from "react";
import styled from "../styled-components";
import { CloseSmallIcon } from "../icon/CloseSmallIcon";
const CloseButton = styled.div.withConfig({ displayName: "CloseButton", componentId: "sc-193k016" }) `
    cursor: pointer;
    line-height: 1;
    color: ${({ theme }) => theme.colors.closeBtn};
`;
const ModalSelectBoxWrapper = styled.div.withConfig({ displayName: "ModalSelectBoxWrapper", componentId: "sc-c2se3z" }) `
    padding: 16px;
    border-radius: 8px;
    background-color: ${props => props.theme.colors.overlayBg};
    border: 1px solid #D0DEF2;
    box-shadow: 0 2px 4px 0 rgba(51,69,100,0.07), 0 6px 16px 0 rgba(51,69,100,0.08);
`;
export class ModalSelectBox extends React.Component {
    render() {
        return (React.createElement(ModalSelectBoxWrapper, null,
            React.createElement("div", { style: { padding: "4px" } },
                React.createElement(CloseButton, { onClick: this.props.onClose },
                    React.createElement(CloseSmallIcon, null))),
            this.props.children));
    }
}
