// @alethio/explorer-ui/lib/qrCode/QrCodeLayer.js
import * as tslib_1 from "tslib";
import * as React from "react";
import styled from "@alethio/ui/lib/styled-components";
import { observable } from "mobx";
import { observer } from "mobx-react";
import { SpinnerRegular } from "@alethio/ui/lib/fx/SpinnerRegular";
const QrCodeLayerRoot = styled.div `
    width: 148px;
    height: 148px;
    position: relative;
`;
const SpinnerWrapper = styled.div `
    position: absolute;
    top: 50%;
    left: 50%;
    transform: translate(-50%, -50%);
`;
let QrCodeLayer = class QrCodeLayer extends React.Component {
    componentDidMount() {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            if (!this.QrCodeComponent) {
                import("./QrCode").then(({ QrCode }) => {
                    this.QrCodeComponent = QrCode;
                }).catch(e => {
                    this.props.logger.error("Couldn't load QrCode component", e);
                });
            }
        });
    }
    render() {
        let QrCodeComponent = this.QrCodeComponent;
        return (React.createElement(QrCodeLayerRoot, null, QrCodeComponent ?
            React.createElement(QrCodeComponent, { value: this.props.value, logger: this.props.logger })
            :
                React.createElement(SpinnerWrapper, null,
                    React.createElement(SpinnerRegular, null))));
    }
};
tslib_1.__decorate([
    observable
], QrCodeLayer.prototype, "QrCodeComponent", void 0);
QrCodeLayer = tslib_1.__decorate([
    observer
], QrCodeLayer);
export { QrCodeLayer };
