// @alethio/ui/lib/layout/content/box/Box.js
import styled, { css } from "../../../styled-components";
import * as React from "react";
const IconRoot = styled.div.withConfig({ displayName: "IconRoot", componentId: "sc-1x8jexb" }) `
    display: block;
`;
const ContentRoot = styled("div").withConfig({ displayName: "ContentRoot", componentId: "sc-1h36b7j" }) `
    padding-top: ${props => props.paddingTop}px;
    white-space: nowrap;

    ${props => props.fullWidth ? css `
    width: 100%;
    ` : ``}
`;
const $Box = ({ children, className, Icon, iconPlacement, metrics, fullWidth }) => {
    let icon = Icon ?
        React.createElement(IconRoot, null,
            React.createElement(Icon, { size: metrics.iconSize }))
        : null;
    let iconIsLeft = iconPlacement === "left";
    return React.createElement("div", { className: className },
        iconIsLeft && icon,
        children !== void 0 ?
            React.createElement(ContentRoot, { fullWidth: fullWidth, paddingTop: metrics.textPaddingTop }, children)
            : null,
        !iconIsLeft && icon);
};
const getColors = (colors, theme) => {
    if (typeof colors === "function") {
        return colors(theme);
    }
    return colors;
};
const BORDER_WIDTH = 1;
export const Box = styled($Box).withConfig({ displayName: "Box", componentId: "sc-p4qq1s" }) `
    display: flex;
    font-size: ${props => props.metrics.fontSize}px;
    line-height: ${props => props.metrics.lineHeight}px;
    height: ${props => props.metrics.height}px;

    ${props => props.fullWidth ? css `
    width: 100%;
    ` : ``}

    padding: 0 ${props => props.metrics.textPaddingX - BORDER_WIDTH}px;

    ${props => {
    if (!props.Icon) {
        return;
    }
    let iconSpacing = Math.max(0, (props.metrics.height - props.metrics.iconSize) / 2 - BORDER_WIDTH);
    let iconIsLeft = props.iconPlacement === "left";
    let hasText = props.children !== void 0;
    return css `
            ${iconIsLeft ? css `
                padding-left: ${iconSpacing}px;
                    ${!hasText ? css `
                    padding-right: ${iconSpacing}px;
                    ` : ``}
                ` : css `
                padding-right: ${iconSpacing}px;
                    ${!hasText ? css `
                    padding-left: ${iconSpacing}px;
                    ` : ``}
            `}

            & ${IconRoot} {
                margin-top: ${iconSpacing}px;

                ${hasText ? css `
                    ${iconIsLeft ? css `
                        margin-right: 8px;
                        ` : css `
                        margin-left: 8px;`}
                ` : ``}
            }
        `;
}}

    font-weight: ${props => props.metrics.fontWeight};
    letter-spacing: ${props => props.metrics.letterSpacing};
    box-sizing: border-box;
    vertical-align: middle;

    border: ${BORDER_WIDTH}px solid ${props => (getColors(props.colors, props.theme).border ||
    getColors(props.colors, props.theme).background ||
    "transparent")};
    background-color: ${props => (getColors(props.colors, props.theme).background ||
    "transparent")};
    color: ${props => getColors(props.colors, props.theme).text};
`;
